File: crates/wicked-estate-extract/tests/fixtures/elixir/sample.ex
defmodule Geometry do
  @moduledoc """
  Basic geometric shape calculations.
  """

  import Enum, only: [max_by: 2, map: 2]
  alias Geometry.Shape, as: Shape
  use Agent

  defstruct [:kind, :params]

  @type t :: %__MODULE__{kind: atom(), params: list(number())}

  @doc "Compute the area of a shape struct."
  def area(%__MODULE__{kind: :circle, params: [r]}), do: :math.pi() * r * r
  def area(%__MODULE__{kind: :rectangle, params: [w, h]}), do: w * h
  def area(%__MODULE__{kind: :triangle, params: [a, b, c]}) do
    s = (a + b + c) / 2.0
    :math.sqrt(s * (s - a) * (s - b) * (s - c))
  end

  @doc "Return the shape with the largest area from a list."
  def largest(shapes) when is_list(shapes) do
    Enum.max_by(shapes, &area/1)
  end

  @doc "Describe a shape as a human-readable string."
  def describe(%__MODULE__{kind: :circle, params: [r]}),
    do: "Circle with radius #{r}"
  def describe(%__MODULE__{kind: :rectangle, params: [w, h]}),
    do: "Rectangle #{w}x#{h}"
  def describe(%__MODULE__{kind: :triangle, params: params}),
    do: "Triangle with sides #{Enum.join(params, ", ")}"

  @doc "Build a circle shape."
  def circle(r), do: %__MODULE__{kind: :circle, params: [r]}

  defp validate_positive(n) when n > 0, do: :ok
  defp validate_positive(_), do: {:error, :non_positive}
end

File: crates/wicked-estate-extract/tests/fixtures/haskell/sample.hs
module Sample where

import Data.List (sort, nub)
import Data.Maybe (fromMaybe, mapMaybe)

-- Type alias
type Name = String
type Score = Double

-- Data type
data Shape
  = Circle Double
  | Rectangle Double Double
  | Triangle Double Double Double
  deriving (Show, Eq)

-- Type class instance
class Describable a where
  describe :: a -> String

instance Describable Shape where
  describe (Circle r)        = "Circle with radius " ++ show r
  describe (Rectangle w h)   = "Rectangle " ++ show w ++ "x" ++ show h
  describe (Triangle a b c)  = "Triangle with sides " ++ show a ++ ", " ++ show b ++ ", " ++ show c

-- Compute the area of a shape
area :: Shape -> Double
area (Circle r)       = pi * r * r
area (Rectangle w h)  = w * h
area (Triangle a b c) =
  let s = (a + b + c) / 2
  in sqrt (s * (s - a) * (s - b) * (s - c))

-- Rank shapes by area descending
rankByArea :: [Shape] -> [Shape]
rankByArea shapes =
  let pairs = map (\s -> (area s, s)) shapes
      sorted = reverse $ sort $ map fst pairs
  in mapMaybe (\score -> lookup score pairs) sorted

-- Score a named entity
scoreEntity :: Name -> Shape -> (Name, Score)
scoreEntity name shape = (name, area shape)
